File: threads/src/main/java/im/threads/internal/formatters/OutgoingMessageCreator.java
package im.threads.internal.formatters;

import android.content.Context;
import android.text.TextUtils;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;

import im.threads.R;
import im.threads.internal.utils.ThreadsLogger;
import im.threads.internal.model.ConsultInfo;
import im.threads.internal.model.FileDescription;
import im.threads.internal.model.Quote;
import im.threads.internal.model.Survey;
import im.threads.internal.model.UserPhrase;
import im.threads.internal.utils.AppInfoHelper;
import im.threads.internal.utils.DeviceInfoHelper;
import im.threads.internal.utils.PrefUtils;

public final class OutgoingMessageCreator {
    private static final String TAG = "MessageFormatter ";

    private static final String ERROR_FORMATTING_JSON = "error formatting json";

    private static String userAgent = "";

    private OutgoingMessageCreator() {
    }

    public static String createUserPhraseMessage(UserPhrase upcomingUserMessage, ConsultInfo consultInfo,
                                                 String quoteMfmsFilePath, String mfmsFilePath,
                                                 String clientId, Long threadId) {
        try {
            Quote quote = upcomingUserMessage.getQuote();
            FileDescription fileDescription = upcomingUserMessage.getFileDescription();
            JSONObject formattedMessage = new JSONObject();
            formattedMessage.put(PushMessageAttributes.UUID, upcomingUserMessage.getUuid());
            formattedMessage.put(PushMessageAttributes.CLIENT_ID, clientId);
            final String phrase = upcomingUserMessage.getPhrase();
            formattedMessage.put(PushMessageAttributes.TEXT, phrase == null ? "" : phrase);
            if (threadId != null && threadId != -1) {
                formattedMessage.put(PushMessageAttributes.THREAD_ID, String.valueOf(threadId));
            }
            formattedMessage.put(PushMessageAttributes.APP_MARKER_KEY, PrefUtils.getAppMarker());
            if (quote != null) {
                JSONArray quotes = new JSONArray();
                formattedMessage.put(PushMessageAttributes.QUOTES, quotes);
                JSONObject quoteJson = new JSONObject();
                quotes.put(quoteJson);
                if (!TextUtils.isEmpty(quote.getText())) {
                    quoteJson.put(PushMessageAttributes.TEXT, quote.getText());
                }
                if (null != consultInfo)
                    quoteJson.put("operator", consultInfo.toJson());//TODO #THREADS-5270 What is it for?
                if (quote.getFileDescription() != null && quoteMfmsFilePath != null) {
                    quoteJson.put(PushMessageAttributes.ATTACHMENTS, attachmentsFromFileDescription(quote.getFileDescription(), quoteMfmsFilePath));
                }
                if (upcomingUserMessage.getQuote().getUuid() != null) {
                    quoteJson.put(PushMessageAttributes.UUID, upcomingUserMessage.getQuote().getUuid());
                }
            }
            if (fileDescription != null && mfmsFilePath != null) {
                formattedMessage.put(PushMessageAttributes.ATTACHMENTS, attachmentsFromFileDescription(fileDescription, mfmsFilePath));
            }
            return formattedMessage.toString();
        } catch (JSONException e) {
            ThreadsLogger.e(TAG, ERROR_FORMATTING_JSON, e);
        }
        return "";
    }

    private static JSONArray attachmentsFromFileDescription(File file) throws JSONException {
        JSONArray attachments = new JSONArray();
        JSONObject attachment = new JSONObject();
        attachments.put(attachment);
        JSONObject optional = new JSONObject();
        attachment.put("optional", optional);
        String extension = file.getAbsolutePath().substring(file.getAbsolutePath().lastIndexOf(".") + 1).toLowerCase();
        String type = null;
        if (extension.equals("jpg")) type = "image/jpg";
        if (extension.equals("png")) type = "image/png";
        if (extension.equals("pdf")) type = "text/pdf";
        optional.put(PushMessageAttributes.TYPE, type);
        optional.put("name", file.getName());
        optional.put("size", file.length());
        optional.put("lastModified", file.lastModified());
        return attachments;
    }

    private static JSONArray attachmentsFromMfmsPath(FileDescription fileDescription) throws JSONException {
        JSONArray attachments = new JSONArray();
        JSONObject attachment = new JSONObject();
        attachments.put(attachment);
        JSONObject optional = new JSONObject();
        attachment.put("optional", optional);
        if (fileDescription.getIncomingName() != null) {
            String extension = fileDescription.getIncomingName().substring(fileDescription.getIncomingName().lastIndexOf(".") + 1).toLowerCase();
            String type = null;
            if (extension.equals("jpg")) type = "image/jpg";
            if (extension.equals("png")) type = "image/png";
            if (extension.equals("pdf")) type = "text/pdf";
            optional.put(PushMessageAttributes.TYPE, type);
        }
        optional.put("name", fileDescription.getIncomingName());
        optional.put("size", fileDescription.getSize());
        optional.put("lastModified", System.currentTimeMillis());
        return attachments;
    }

    private static JSONArray attachmentsFromFileDescription(FileDescription fileDescription, String mfmsFilepath) throws JSONException {
        JSONArray attachments = null;
        if (fileDescription.getFilePath() != null && new File(fileDescription.getFilePath()).exists()) {
            attachments = attachmentsFromFileDescription(new File(fileDescription.getFilePath()));
        } else if (fileDescription.getDownloadPath() != null) {
            attachments = attachmentsFromMfmsPath(fileDescription);
        }
        if (attachments != null) {
            attachments.getJSONObject(0).put("result", mfmsFilepath);
        }
        return attachments;
    }

    public static String createEnvironmentMessage(String clientName, String clientId, boolean clientIdEncrypted, String data, Context ctx) {
        JSONObject object = new JSONObject();
        try {
            object.put("name", clientName);
            object.put(PushMessageAttributes.CLIENT_ID, clientId);
            object.put(PushMessageAttributes.CLIENT_ID_ENCRYPTED, clientIdEncrypted);
            object.put(PushMessageAttributes.DATA, data);
            object.put("platform", "Android");
            object.put("osVersion", DeviceInfoHelper.getOsVersion());
            object.put("device", DeviceInfoHelper.getDeviceName());
            object.put("ip", DeviceInfoHelper.getIpAddress());
            object.put("appVersion", AppInfoHelper.getAppVersion(ctx));
            object.put("appName", AppInfoHelper.getAppName(ctx));
            object.put(PushMessageAttributes.APP_BUNDLE_KEY, AppInfoHelper.getAppId(ctx));
            object.put(PushMessageAttributes.APP_MARKER_KEY, PrefUtils.getAppMarker());
            object.put("libVersion", AppInfoHelper.getLibVersion());
            object.put("clientLocale", DeviceInfoHelper.getLocale(ctx));
            object.put(PushMessageAttributes.TYPE, PushMessageType.CLIENT_INFO.name());
        } catch (JSONException e) {
            ThreadsLogger.e(TAG, "createEnvironmentMessage", e);
        }
        return object.toString();
    }

    public static String createRatingDoneMessage(Survey survey, String clientId, String appMarker) {
        JSONObject object = new JSONObject();
        try {
            object.put(PushMessageAttributes.CLIENT_ID, clientId);
            object.put(PushMessageAttributes.TYPE, PushMessageType.SURVEY_QUESTION_ANSWER.name());
            object.put("sendingId", survey.getSendingId());
            object.put("questionId", survey.getQuestions().get(0).getId());
            object.put("rate", survey.getQuestions().get(0).getRate());
            object.put("text", survey.getQuestions().get(0).getText());
            object.put(PushMessageAttributes.APP_MARKER_KEY, appMarker);
        } catch (JSONException e) {
            ThreadsLogger.e(TAG, "createRatingDoneMessage", e);
        }
        return object.toString();
    }

    public static String createRatingReceivedMessage(long sendingId, String clientId) {
        JSONObject object = new JSONObject();
        try {
            object.put(PushMessageAttributes.CLIENT_ID, clientId);
            object.put(PushMessageAttributes.TYPE, PushMessageType.SURVEY_PASSED.name());
            object.put(PushMessageAttributes.APP_MARKER_KEY, PrefUtils.getAppMarker());
            object.put("sendingId", sendingId);
        } catch (JSONException e) {
            ThreadsLogger.e(TAG, "createRatingReceivedMessage", e);
        }
        return object.toString();
    }

    public static String createResolveThreadMessage(String clientId) {
        JSONObject object = new JSONObject();
        try {
            object.put(PushMessageAttributes.CLIENT_ID, clientId);
            object.put(PushMessageAttributes.TYPE, PushMessageType.CLOSE_THREAD.name());
            object.put(PushMessageAttributes.APP_MARKER_KEY, PrefUtils.getAppMarker());
        } catch (JSONException e) {
            ThreadsLogger.e(TAG, "createResolveThreadMessage", e);
        }

        return object.toString();
    }

    public static String createReopenThreadMessage(String clientId) {
        JSONObject object = new JSONObject();
        try {
            object.put(PushMessageAttributes.CLIENT_ID, clientId);
            object.put(PushMessageAttributes.TYPE, PushMessageType.REOPEN_THREAD.name());
            object.put(PushMessageAttributes.APP_MARKER_KEY, PrefUtils.getAppMarker());
        } catch (JSONException e) {
            ThreadsLogger.e(TAG, "createReopenThreadMessage", e);
        }

        return object.toString();
    }

    public static String createMessageTyping(String clientId, String input) {
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put(PushMessageAttributes.CLIENT_ID, clientId);
            jsonObject.put(PushMessageAttributes.TYPE, PushMessageType.TYPING.name());
            jsonObject.put(PushMessageAttributes.TYPING_DRAFT, input);
            jsonObject.put(PushMessageAttributes.APP_MARKER_KEY, PrefUtils.getAppMarker());
        } catch (JSONException e) {
            ThreadsLogger.e(TAG, "createMessageTyping", e);
        }
        return jsonObject.toString().replace("\\\\", "");
    }

    public static String createMessageClientOffline(String clientId) {
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put(PushMessageAttributes.CLIENT_ID, clientId);
            jsonObject.put(PushMessageAttributes.TYPE, PushMessageType.CLIENT_OFFLINE.name());
            jsonObject.put(PushMessageAttributes.APP_MARKER_KEY, PrefUtils.getAppMarker());
        } catch (JSONException e) {
            ThreadsLogger.e(TAG, "createMessageClientOffline", e);
        }
        return jsonObject.toString().replace("\\\\", "");
    }

    public static String getUserAgent(Context ctx) {
        if (TextUtils.isEmpty(userAgent)) {
            userAgent = String.format(ctx.getResources().getString(R.string.threads_user_agent),
                    DeviceInfoHelper.getOsVersion(),
                    DeviceInfoHelper.getDeviceName(),
                    DeviceInfoHelper.getIpAddress(),
                    AppInfoHelper.getAppVersion(ctx),
                    AppInfoHelper.getAppId(ctx),
                    AppInfoHelper.getLibVersion());
        }
        return userAgent;
    }

    public static String createInitChatMessage(String clientId, String data) {
        JSONObject object = new JSONObject();
        try {
            object.put(PushMessageAttributes.CLIENT_ID, clientId);
            object.put(PushMessageAttributes.TYPE, PushMessageType.INIT_CHAT.name());
            object.put(PushMessageAttributes.DATA, data);
            object.put(PushMessageAttributes.APP_MARKER_KEY, PrefUtils.getAppMarker());
        } catch (JSONException e) {
            ThreadsLogger.e(TAG, "createInitChatMessage", e);
        }
        return object.toString();
    }
}

File: threads/src/main/java/im/threads/business/transport/OutgoingMessageCreator.kt
package im.threads.business.transport

import android.net.Uri
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import im.threads.business.UserInfoBuilder
import im.threads.business.formatters.ChatItemType
import im.threads.business.models.CAMPAIGN_DATE_FORMAT
import im.threads.business.models.ConsultInfo
import im.threads.business.models.FileDescription
import im.threads.business.models.Survey
import im.threads.business.models.UserPhrase
import im.threads.business.preferences.Preferences
import im.threads.business.preferences.PreferencesCoreKeys
import im.threads.business.rest.queries.ThreadsApi
import im.threads.business.utils.AppInfo
import im.threads.business.utils.ClientUseCase
import im.threads.business.utils.DeviceInfo
import im.threads.business.utils.FileUtils.getFileName
import im.threads.business.utils.FileUtils.getFileSize
import im.threads.business.utils.FileUtils.getMimeType
import java.text.SimpleDateFormat
import java.util.Locale
import java.util.TimeZone

class OutgoingMessageCreator(
    private val preferences: Preferences,
    private val clientUseCase: ClientUseCase,
    private val appInfo: AppInfo,
    private val deviceInfo: DeviceInfo
) {
    fun createInitChatMessage(userInfo: UserInfoBuilder): JsonObject {
        val jsonObject = JsonObject()
        val deviceAddress = preferences.get<String>(PreferencesCoreKeys.DEVICE_ADDRESS)
        jsonObject.apply {
            addProperty(MessageAttributes.CLIENT_ID, userInfo.clientId)
            addProperty(MessageAttributes.TYPE, ChatItemType.INIT_CHAT.name)
            addProperty(MessageAttributes.DATA, userInfo.clientData)
            addProperty(MessageAttributes.APP_MARKER_KEY, userInfo.appMarker)
            addProperty(MessageAttributes.DEVICE_ADDRESS, deviceAddress)
        }
        return jsonObject
    }

    fun createClientInfoMessage(userInfo: UserInfoBuilder, locale: String, isPreregister: Boolean): JsonObject {
        val deviceAddress = preferences.get(PreferencesCoreKeys.DEVICE_ADDRESS, "")
        val jsonObject = JsonObject().apply {
            addProperty("name", userInfo.userName)
            addProperty(MessageAttributes.CLIENT_ID, userInfo.clientId)
            addProperty(MessageAttributes.CLIENT_ID_ENCRYPTED, userInfo.clientIdEncrypted)
            userInfo.clientData?.let { addProperty(MessageAttributes.DATA, it) }
            addProperty("platform", "Android")
            addProperty("osVersion", deviceInfo.osVersion)
            addProperty("device", deviceInfo.deviceName)
            addProperty("ip", deviceInfo.ipAddress)
            addProperty("appVersion", appInfo.appVersion)
            addProperty("appName", appInfo.appName)
            addProperty(MessageAttributes.APP_BUNDLE_KEY, appInfo.appId)
            addProperty(MessageAttributes.APP_MARKER_KEY, userInfo.appMarker)
            addProperty("libVersion", appInfo.libVersion)
            addProperty("clientLocale", locale)
            addProperty("chatApiVersion", ThreadsApi.getApiVersion())
            addProperty(MessageAttributes.TYPE, ChatItemType.CLIENT_INFO.name)
            addProperty(MessageAttributes.DEVICE_ADDRESS, deviceAddress)
            addProperty(MessageAttributes.AUTHORIZED, true)
            if (isPreregister) {
                addProperty("preRegister", true)
            }
        }
        return jsonObject
    }

    fun createMessageTyping(userInfo: UserInfoBuilder, input: String?): JsonObject {
        val deviceAddress = preferences.get(PreferencesCoreKeys.DEVICE_ADDRESS) ?: ""
        val jsonObject = JsonObject().apply {
            addProperty(MessageAttributes.TYPE, ChatItemType.TYPING.name)
            addProperty(MessageAttributes.TYPING_DRAFT, input)
            addProperty(MessageAttributes.APP_MARKER_KEY, userInfo.appMarker)
            addProperty(MessageAttributes.DEVICE_ADDRESS, deviceAddress)
        }
        return jsonObject
    }

    fun createRatingDoneMessage(userInfo: UserInfoBuilder, survey: Survey): JsonObject {
        val jsonObject = JsonObject().apply {
            addProperty(MessageAttributes.CLIENT_ID, userInfo.clientId)
            addProperty(MessageAttributes.TYPE, ChatItemType.SURVEY_QUESTION_ANSWER.name)
            addProperty("sendingId", survey.sendingId)
            if (!survey.questions.isNullOrEmpty()) {
                addProperty("questionId", survey.questions!![0].id)
                addProperty("rate", survey.questions!![0].rate)
                addProperty("text", survey.questions!![0].text)
            }
            addProperty(MessageAttributes.APP_MARKER_KEY, userInfo.appMarker)
        }
        return jsonObject
    }

    fun createResolveThreadMessage(userInfo: UserInfoBuilder): JsonObject {
        val jsonObject = JsonObject().apply {
            addProperty(MessageAttributes.CLIENT_ID, userInfo.clientId)
            addProperty(MessageAttributes.TYPE, ChatItemType.CLOSE_THREAD.name)
            addProperty(MessageAttributes.APP_MARKER_KEY, userInfo.appMarker)
        }
        return jsonObject
    }

    fun createReopenThreadMessage(userInfo: UserInfoBuilder): JsonObject {
        val jsonObject = JsonObject().apply {
            addProperty(MessageAttributes.CLIENT_ID, userInfo.clientId)
            addProperty(MessageAttributes.TYPE, ChatItemType.REOPEN_THREAD.name)
            addProperty(MessageAttributes.APP_MARKER_KEY, userInfo.appMarker)
        }
        return jsonObject
    }

    fun createMessageClientOffline(clientId: String, appMarker: String?): JsonObject {
        val jsonObject = JsonObject().apply {
            addProperty(MessageAttributes.CLIENT_ID, clientId)
            addProperty(MessageAttributes.TYPE, ChatItemType.CLIENT_OFFLINE.name)
            addProperty(MessageAttributes.APP_MARKER_KEY, appMarker)
        }
        return jsonObject
    }

    fun createMessageUpdateLocation(
        latitude: Double,
        longitude: Double,
        locale: String,
        userInfo: UserInfoBuilder
    ): JsonObject {
        val jsonObject = JsonObject().apply {
            val location = JsonObject().apply {
                addProperty(MessageAttributes.COORDINATES, "$latitude, $longitude")
            }
            addProperty(MessageAttributes.DATA, location.toString())
            addProperty("name", userInfo.userName)
            addProperty(MessageAttributes.CLIENT_ID, userInfo.clientId)
            addProperty(MessageAttributes.CLIENT_ID_ENCRYPTED, userInfo.clientIdEncrypted)
            addProperty("platform", "Android")
            addProperty("osVersion", deviceInfo.osVersion)
            addProperty("device", deviceInfo.deviceName)
            addProperty("ip", deviceInfo.ipAddress)
            addProperty("appVersion", appInfo.appVersion)
            addProperty("appName", appInfo.appName)
            addProperty(MessageAttributes.APP_BUNDLE_KEY, appInfo.appId)
            addProperty(MessageAttributes.APP_MARKER_KEY, userInfo.appMarker)
            addProperty("libVersion", appInfo.libVersion)
            addProperty("clientLocale", locale)
            addProperty("chatApiVersion", ThreadsApi.getApiVersion())
            addProperty(MessageAttributes.TYPE, ChatItemType.UPDATE_LOCATION.name)
        }
        return jsonObject
    }

    fun createUserPhraseMessage(
        userPhrase: UserPhrase,
        consultInfo: ConsultInfo?,
        quoteMfmsFilePath: String?,
        mfmsFilePath: String?,
        userInfo: UserInfoBuilder
    ): JsonObject {
        val phrase = userPhrase.phraseText
        val quote = userPhrase.quote
        val fileDescription = userPhrase.fileDescription
        val campaignMessage = userPhrase.campaignMessage
        val deviceAddress = preferences.get(PreferencesCoreKeys.DEVICE_ADDRESS) ?: ""
        val formattedMessage = JsonObject().apply {
            addProperty(MessageAttributes.UUID, userPhrase.id)
            addProperty(MessageAttributes.CLIENT_ID, userInfo.clientId)
            addProperty(MessageAttributes.TEXT, phrase ?: "")
            addProperty(MessageAttributes.APP_MARKER_KEY, userInfo.appMarker)
            addProperty(MessageAttributes.DEVICE_ADDRESS, deviceAddress)
            addProperty(MessageAttributes.TYPE, ChatItemType.MESSAGE.name)
        }
        val quotes = JsonArray().apply {
            campaignMessage?.let {
                val sdf = SimpleDateFormat(CAMPAIGN_DATE_FORMAT, Locale.getDefault())
                sdf.timeZone = TimeZone.getTimeZone("UTC")
                val quoteJson = JsonObject().apply {
                    addProperty(MessageAttributes.UUID, it.chatMessageId)
                    addProperty(MessageAttributes.TEXT, it.text)
                    addProperty(MessageAttributes.IS_MASS_PUSH_MESSAGE, true)
                    addProperty(MessageAttributes.CAMPAIGN, it.campaign)
                    addProperty(
                        MessageAttributes.RECEIVED_DATE,
                        sdf.format(it.receivedDate)
                    )
                }
                add(quoteJson)

                val routingParams = JsonObject().apply {
                    addProperty(MessageAttributes.PRIORITY, it.priority)
                    addProperty(MessageAttributes.SKILL_ID, it.skillId)
                    addProperty(MessageAttributes.EXPIRED_AT, sdf.format(it.expiredAt))
                }
                formattedMessage.add(MessageAttributes.ROUTING_PARAMS, routingParams)
            }
        }

        quote?.let {
            val quoteJson = JsonObject().apply {
                if (!it.text.isNullOrEmpty()) {
                    addProperty(MessageAttributes.TEXT, it.text)
                }
                consultInfo?.let {
                    add("operator", it.toJson())
                }
                it.fileDescription?.let {
                    if (!quoteMfmsFilePath.isNullOrEmpty()) {
                        add(
                            MessageAttributes.ATTACHMENTS,
                            attachmentsFromFileDescription(it, quoteMfmsFilePath)
                        )
                    }
                }
                if (!userPhrase.quote.uuid.isNullOrEmpty()) {
                    addProperty(MessageAttributes.UUID, userPhrase.quote.uuid)
                }
            }
            quotes.add(quoteJson)
        }
        formattedMessage.add(MessageAttributes.QUOTES, quotes)

        fileDescription?.let {
            if (!mfmsFilePath.isNullOrEmpty()) {
                formattedMessage.add(
                    MessageAttributes.ATTACHMENTS,
                    attachmentsFromFileDescription(fileDescription, mfmsFilePath)
                )
            }
        }
        return formattedMessage
    }

    private fun attachmentsFromFileDescription(
        fileDescription: FileDescription?
    ): JsonArray {
        val attachments = JsonArray()
        fileDescription?.let {
            val optional = JsonObject().apply {
                addProperty(MessageAttributes.TYPE, getMimeType(it))
                val fileUri = fileDescription.fileUri
                fileUri?.let {
                    addProperty("name", getFileName(it))
                    addProperty("size", getFileSize(it))
                }
                addProperty("lastModified", 0)
            }
            val attachment = JsonObject().apply {
                add("optional", optional)
            }
            attachments.add(attachment)
        }
        return attachments
    }

    private fun attachmentsFromMfmsPath(fileDescription: FileDescription?): JsonArray {
        val attachments = JsonArray()
        fileDescription?.let {
            val optional = JsonObject().apply {
                if (!it.incomingName.isNullOrEmpty()) {
                    addProperty(MessageAttributes.TYPE, getMimeType(Uri.parse(it.incomingName)))
                }
                addProperty("name", it.incomingName)
                addProperty("size", it.size)
                addProperty("lastModified", System.currentTimeMillis())
            }
            val attachment = JsonObject().apply {
                add("optional", optional)
            }
            attachments.add(attachment)
        }
        return attachments
    }

    private fun attachmentsFromFileDescription(
        fileDescription: FileDescription?,
        mfmsFilepath: String
    ): JsonArray? {
        var attachments: JsonArray? = null

        fileDescription?.let {
            val file = it
            it.fileUri?.let {
                attachments = attachmentsFromFileDescription(file)
            }
            it.downloadPath?.let {
                attachments = attachmentsFromMfmsPath(file)
            }
        }
        attachments?.let {
            (it[0] as JsonObject).addProperty("result", mfmsFilepath)
        }
        return attachments
    }
}